### PR TITLE
Update JsonToStructs and ScanJson to have white space normalization

### PR DESCRIPTION
**Pull request by revans2**
This also contributes to https://github.com/NVIDIA/spark-rapids/issues/10491 in a very small way by adding in a few more tests.

Mostly it turns on white space normalization and tries to verify that it is doing the right thing, but there were some errors, so I filed more issues.

### CODE REVIEW

**Comment by revans2** [comment]
build

**Comment by revans2** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/ColumnCastUtil.scala [hunk 1]
```diff
@@ -45,12 +45,18 @@ object ColumnCastUtil {
...
+   *                              non-nested type is expected.
    * @param convert the partial function used to convert the data. If this matches and returns
    *                a updated view this function takes ownership of that view.
    * @return None if there were no changes to the view or the updated view along with anything else
    *         that needs to be closed.
    */
-  def deepTransformView(cv: ColumnView, dt: Option[DataType] = None)
+  def deepTransformView(cv: ColumnView, dt: Option[DataType] = None,
+      nestedMismatchHandler: Option[(ColumnView, DataType) =>
+          (Option[ColumnView], ArrayBuffer[AutoCloseable])] = None)
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/ColumnCastUtil.scala:59]: Does the handler need to return a mutable ArrayBuffer?  I think the handler could return an immutable Seq given how it's being used, and that seems more flexible and less error-prone than forcing an ArrayBuffer here.

**Review by jlowe — APPROVED** [review]
Minor comment, overall lgtm.

**Comment by revans2** [comment]
build

**Comment by revans2** [comment]
@jlowe please take another look

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonReadCommon.scala [hunk 2]
```diff
@@ -266,10 +268,53 @@ object GpuJsonReadCommon {
   private def timestampFormat(options: JSONOptions): String =
     GpuJsonUtils.timestampFormatInRead(options)
 
+  private def throwMismatchException(cv: ColumnView,
+      dt: DataType): (Option[ColumnView], ArrayBuffer[AutoCloseable]) = {
+    throw new IllegalStateException(s"Don't know how to transform $cv to $dt for JSON")
+  }
+
+  private def nestedColumnViewMismatchTransform(cv: ColumnView,
+      dt: DataType): (Option[ColumnView], ArrayBuffer[AutoCloseable]) = {
```
Comment (jlowe) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonReadCommon.scala:277]:
```suggestion
      dt: DataType): (Option[ColumnView], Seq[AutoCloseable]) = {
```

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonReadCommon.scala [hunk 3]
```diff
@@ -266,10 +268,53 @@ object GpuJsonReadCommon {
...
+        dt match {
+          case ByteType | ShortType | IntegerType | LongType |
+               BooleanType | FloatType | DoubleType |
+               _: DecimalType | _: StructType =>
+            // This is all nulls
+            val rows = cv.getRowCount().toInt
+            val ret = withResource(GpuScalar.from(null, dt)) { nullScalar =>
+              ColumnVector.fromScalar(nullScalar, rows)
+            }
+            (Some(ret.asInstanceOf[ColumnView]), ArrayBuffer(ret))
```
Comment (jlowe) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonReadCommon.scala:292]:
```suggestion
            (Some(ret.asInstanceOf[ColumnView]), Seq(ret))
```

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonReadCommon.scala [hunk 4]
```diff
@@ -266,10 +268,53 @@ object GpuJsonReadCommon {
...
+        }
+      case DType.STRUCT =>
+        dt match {
+          case _: ArrayType =>
+            // This is all nulls
+            val rows = cv.getRowCount().toInt
+            val ret = withResource(GpuScalar.from(null, dt)) { nullScalar =>
+              ColumnVector.fromScalar(nullScalar, rows)
+            }
+            (Some(ret.asInstanceOf[ColumnView]), ArrayBuffer(ret))
```
Comment (jlowe) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonReadCommon.scala:304]:
```suggestion
            (Some(ret.asInstanceOf[ColumnView]), Seq(ret))
```

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonReadCommon.scala [hunk 5]
```diff
@@ -266,10 +268,53 @@ object GpuJsonReadCommon {
   private def timestampFormat(options: JSONOptions): String =
     GpuJsonUtils.timestampFormatInRead(options)
 
+  private def throwMismatchException(cv: ColumnView,
+      dt: DataType): (Option[ColumnView], ArrayBuffer[AutoCloseable]) = {
```
Comment (jlowe) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonReadCommon.scala:272]:
```suggestion
      dt: DataType): (Option[ColumnView], Seq[AutoCloseable]) = {
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/ColumnCastUtil.scala [hunk 6]
```diff
@@ -143,13 +168,19 @@ object ColumnCastUtil {
...
+   *                              needed, or when a nested value is returned by CUDF but a
+   *                              non-nested type is expected.
    * @param convert the partial function used to convert the data. If this matches and returns
    *                a updated view this function takes ownership of that view.
    * @return the updated vector
    */
-  def deepTransform(cv: ColumnVector, dt: Option[DataType] = None)
+  def deepTransform(cv: ColumnVector, dt: Option[DataType] = None,
+      nestedMismatchHandler: Option[(ColumnView, DataType) =>
+          (Option[ColumnView], ArrayBuffer[AutoCloseable])] = None)
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/ColumnCastUtil.scala:181]:
```suggestion
          (Option[ColumnView], Seq[AutoCloseable])] = None)
```

**Review by jlowe — COMMENTED** [review]
Looks like the ArrayBuffer -> Seq change only happened halfway, suggested some updates.

**Comment by revans2** [comment]
build

**Comment by revans2** [comment]
@jlowe please take another look

**Review by jlowe — APPROVED** [review]

**Pull request merged**